File: feign-core/src/main/java/name/abhijitsarkar/feign/core/matcher/DefaultMethodMatcher.java
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * License for more details.
 */

package name.abhijitsarkar.feign.core.matcher;

import lombok.extern.slf4j.Slf4j;
import name.abhijitsarkar.feign.model.Request;
import name.abhijitsarkar.feign.model.FeignMapping;
import name.abhijitsarkar.feign.model.RequestProperties;

import java.util.Locale;
import java.util.function.BiFunction;

/**
 * @author Abhijit Sarkar
 */
@Slf4j
public class DefaultMethodMatcher implements BiFunction<Request, FeignMapping, Boolean> {
    @Override
    public Boolean apply(Request request, FeignMapping feignMapping) {
        RequestProperties requestProperties = feignMapping.getRequest();
        String requestMethod = request.getMethod();

        String method = requestProperties.getMethod().getName();
        boolean ignoreCase = requestProperties.getMethod().isIgnoreCase();

        boolean match = ignoreCase ? requestMethod.toLowerCase(Locale.ENGLISH).matches(method.toLowerCase(Locale.ENGLISH)) :
                requestMethod.matches(method);

        log.info("Comparing request method: {} with: {}.", requestMethod, method);
        log.info("Method match returned {}.", match);

        return match;
    }
}

File: feign-core/src/main/scala/name/abhijitsarkar/feign/core/matcher/DefaultMethodMatcher.scala
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 */

package name.abhijitsarkar.feign.core.matcher

import java.util.function.BiFunction

import name.abhijitsarkar.feign.Request
import name.abhijitsarkar.feign.core.model.FeignMapping
import org.slf4j.LoggerFactory

/**
  * @author Abhijit Sarkar
  */
class DefaultMethodMatcher extends BiFunction[Request, FeignMapping, Boolean] {
  val logger = LoggerFactory.getLogger(classOf[DefaultMethodMatcher])

  override def apply(request: Request, feignMapping: FeignMapping) = {
    val requestProperties = feignMapping.request
    val requestMethod = request.method

    val method = requestProperties.method.name
    val ignoreCase = requestProperties.method.ignoreCaseToScala.getOrElse(false)

    val maybeToLower = (s: String) => if (ignoreCase) s.toLowerCase() else s

    val m = Some(requestMethod).map(maybeToLower).zip(Some(method).map(maybeToLower)).exists { pair =>
      pair._1.matches(pair._2)
    }

    logger.info(s"Comparing request method: ${requestMethod} with: ${method}.")
    logger.info(s"Method match returned ${m}.")

    m
  }
}
